File: backend/dormammu/state/session_manager.py
"""Session identity, listing, migration, and snapshot operations.

This module owns everything related to *which* session is active, how session
IDs are generated, how old sessions are listed, and how a legacy flat-root
``.dev/`` layout is migrated into the session-subdirectory model.

It is intentionally free of bootstrap-generation and operator-sync concerns.
"""
from __future__ import annotations

import json
import re
import shutil
from pathlib import Path
from typing import Any

from dormammu._utils import iso_now as _iso_now
from dormammu.config import AppConfig
from dormammu.state.persistence import read_json, write_json


_CORE_STATE_FILENAMES = (
    "DASHBOARD.md",
    "PLAN.md",
    "session.json",
    "workflow_state.json",
)
_OPTIONAL_STATE_FILENAMES = (
    "supervisor_report.md",
    "continuation_prompt.txt",
)


class SessionManager:
    """Manage session identity, enumeration, and lifecycle snapshots.

    Parameters
    ----------
    config:
        Runtime application config — used for ``app_name`` and path resolution.
    base_dev_dir:
        The root ``.dev/`` directory (not a session subdirectory).
    sessions_dir:
        The ``.dev/sessions/`` directory that holds per-session subdirectories.
    """

    def __init__(
        self,
        config: AppConfig,
        base_dev_dir: Path,
        sessions_dir: Path,
    ) -> None:
        self.config = config
        self.base_dev_dir = base_dev_dir
        self.sessions_dir = sessions_dir

    # ------------------------------------------------------------------
    # Session ID helpers
    # ------------------------------------------------------------------

    @staticmethod
    def normalize_session_id(value: str) -> str:
        """Return a filesystem-safe session ID from *value*."""
        normalized = re.sub(r"[^a-zA-Z0-9._-]+", "-", value.strip()).strip("-._")
        if not normalized:
            raise ValueError("session_id must contain at least one safe filename character.")
        return normalized

    def generated_session_id(self, timestamp: str) -> str:
        """Generate a unique session ID based on *timestamp*."""
        compact = (
            timestamp.replace("-", "").replace(":", "").replace("+", "-").replace("T", "-")
        )
        base = f"{self.config.app_name}-{compact}"
        candidate = base
        sequence = 1
        while (self.sessions_dir / candidate).exists():
            candidate = f"{base}-{sequence:02d}"
            sequence += 1
        return candidate

    @staticmethod
    def current_session_id(session_path: Path) -> str | None:
        """Read the active session ID recorded inside *session_path*."""
        if not session_path.exists():
            return None
        try:
            payload = read_json(session_path)
        except json.JSONDecodeError:
            return None
        session_id = payload.get("session_id") or payload.get("active_session_id")
        return str(session_id) if session_id else None

    def read_active_session_id(self) -> str | None:
        """Return the active session ID recorded in the root ``.dev/session.json``."""
        return self.current_session_id(self.base_dev_dir / "session.json")

    # ------------------------------------------------------------------
    # Session listing
    # ------------------------------------------------------------------

    def list_sessions(self) -> list[dict[str, Any]]:
        """Return a sorted list of all known sessions with summary metadata."""
        active_session_id = self.read_active_session_id()
        if not self.sessions_dir.exists():
            return []

        sessions: list[dict[str, Any]] = []
        for session_dir in sorted(self.sessions_dir.iterdir()):
            if not session_dir.is_dir():
                continue
            session_path = session_dir / "session.json"
            if not session_path.exists():
                continue

            session_state = read_json(session_path)
            workflow_path = session_dir / "workflow_state.json"
            workflow_state = read_json(workflow_path) if workflow_path.exists() else {}
            bootstrap = session_state.get("bootstrap") or {}
            raw_goal = bootstrap.get("goal") or session_state.get("goal") or ""
            goal_summary = (raw_goal[:120] + "...") if len(raw_goal) > 120 else raw_goal
            loop_state = session_state.get("loop") or workflow_state.get("loop") or {}
            supervisor_verdict = loop_state.get("latest_supervisor_verdict")
            attempts_completed = loop_state.get("attempts_completed")
            session_id = session_state.get("session_id")
            sessions.append(
                {
                    "session_id": session_id,
                    "snapshot_dir": str(session_dir),
                    "created_at": session_state.get("created_at"),
                    "updated_at": session_state.get("updated_at"),
                    "goal": goal_summary,
                    "is_active": session_id == active_session_id,
                    "supervisor_verdict": supervisor_verdict,
                    "attempts_completed": attempts_completed,
                }
            )

        return sorted(
            sessions,
            key=lambda s: (s.get("updated_at") or s.get("created_at") or ""),
            reverse=False,
        )

    # ------------------------------------------------------------------
    # Legacy migration and snapshots
    # ------------------------------------------------------------------

    def has_legacy_root_snapshot(self) -> bool:
        """Return True if the root ``.dev/`` still contains un-migrated state files."""
        return any(
            (self.base_dev_dir / filename).exists()
            for filename in (*_CORE_STATE_FILENAMES, "TASKS.md")
        )

    @staticmethod
    def copy_state_snapshot(source_dir: Path, target_dir: Path) -> None:
        """Copy core and optional state files from *source_dir* to *target_dir*."""
        target_dir.mkdir(parents=True, exist_ok=True)
        for filename in (*_CORE_STATE_FILENAMES, *_OPTIONAL_STATE_FILENAMES):
            source = source_dir / filename
            target = target_dir / filename
            if source.exists():
                target.write_text(source.read_text(encoding="utf-8"), encoding="utf-8")
            elif target.exists():
                target.unlink()

        source_tasks_path = source_dir / "TASKS.md"
        target_tasks_path = target_dir / "TASKS.md"
        if source_tasks_path.exists():
            target_tasks_path.write_text(
                source_tasks_path.read_text(encoding="utf-8"),
                encoding="utf-8",
            )
        elif target_tasks_path.exists():
            target_tasks_path.unlink()

        legacy_tasks_path = source_dir / "TASKS.md"
        target_plan_path = target_dir / "PLAN.md"
        if legacy_tasks_path.exists() and not target_plan_path.exists():
            target_plan_path.write_text(
                legacy_tasks_path.read_text(encoding="utf-8"), encoding="utf-8"
            )

    def migrate_legacy_root_snapshot(self, *, timestamp: str | None = None) -> str | None:
        """Move a flat-root ``.dev/`` layout into a session subdirectory.

        Returns the session ID that was migrated (or already existed), or
        ``None`` when no migration was needed.
        """
        from dormammu.state.models import STATE_SCHEMA_VERSION

        legacy_session_id = self.read_active_session_id()
        if legacy_session_id is not None and (self.sessions_dir / legacy_session_id).exists():
            return legacy_session_id
        if not self.has_legacy_root_snapshot():
            return None

        session_id = legacy_session_id or self.generated_session_id(timestamp or _iso_now())
        target_dir = self.sessions_dir / session_id
        self.copy_state_snapshot(self.base_dev_dir, target_dir)
        legacy_logs_dir = self.base_dev_dir / "logs"
        target_logs_dir = target_dir / "logs"
        if legacy_logs_dir.exists() and not target_logs_dir.exists():
            shutil.copytree(legacy_logs_dir, target_logs_dir)

        session_path = target_dir / "session.json"
        if session_path.exists():
            session_state = read_json(session_path)
            session_state["session_id"] = session_id
            session_state.setdefault("state_schema_version", STATE_SCHEMA_VERSION)
            write_json(session_path, session_state)
        return session_id

    def read_legacy_root_session_payload(self) -> dict[str, Any] | None:
        """Return the root session.json payload if it predates the session model.

        Returns ``None`` when the root session file is already an index
        (contains ``active_session_id``) or does not exist.
        """
        session_path = self.base_dev_dir / "session.json"
        if not session_path.exists():
            return None
        payload = read_json(session_path)
        if "active_session_id" in payload:
            return None
        if "session_id" not in payload:
            return None
        return payload

File: runtime/src/state/sessionManager.ts
import { cp, mkdir, readdir, readFile, stat, unlink, writeFile } from "node:fs/promises";
import path from "node:path";

import { STATE_SCHEMA_VERSION } from "./models.js";
import { readJson, writeJson, type JsonObject } from "./persistence.js";

const CORE_STATE_FILENAMES = [
  "DASHBOARD.md",
  "PLAN.md",
  "session.json",
  "workflow_state.json"
] as const;

const OPTIONAL_STATE_FILENAMES = [
  "supervisor_report.md",
  "continuation_prompt.txt"
] as const;

export type SessionManagerConfig = {
  appName?: string;
};

export type SessionSummary = {
  session_id: unknown;
  snapshot_dir: string;
  created_at: unknown;
  updated_at: unknown;
  goal: string;
  is_active: boolean;
  supervisor_verdict: unknown;
  attempts_completed: unknown;
  active_worktree_id: string | null;
  managed_worktree_count: number;
};

type WorktreeSummary = {
  activeWorktreeId: string | null;
  managedCount: number;
};

async function exists(filePath: string): Promise<boolean> {
  try {
    await stat(filePath);
    return true;
  } catch (error) {
    if (
      error instanceof Error &&
      "code" in error &&
      (error as NodeJS.ErrnoException).code === "ENOENT"
    ) {
      return false;
    }
    throw error;
  }
}

function isRecord(value: unknown): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function normalizeWorktrees(payload: unknown): WorktreeSummary {
  if (!isRecord(payload)) {
    return { activeWorktreeId: null, managedCount: 0 };
  }
  const managedPayload = payload.managed;
  const managedIds = Array.isArray(managedPayload)
    ? managedPayload
        .filter((item): item is JsonObject => isRecord(item))
        .map((item) => item.worktree_id)
        .filter((value): value is string | number | boolean => value != null)
        .map((value) => String(value).trim())
        .filter(Boolean)
    : [];
  const ids = new Set(managedIds);
  const rawActiveId = payload.active_worktree_id;
  const activeWorktreeId =
    rawActiveId == null || !ids.has(String(rawActiveId).trim())
      ? null
      : String(rawActiveId).trim();
  return {
    activeWorktreeId,
    managedCount: ids.size
  };
}

async function copyTextFileIfPresent(source: string, target: string): Promise<void> {
  if (await exists(source)) {
    await writeFile(target, await readFile(source, "utf8"), "utf8");
  } else if (await exists(target)) {
    await unlink(target);
  }
}

export class SessionManager {
  readonly config: Required<SessionManagerConfig>;
  readonly baseDevDir: string;
  readonly sessionsDir: string;
  readonly legacyBaseDevDir: string;

  constructor(options: {
    config?: SessionManagerConfig;
    baseDevDir: string;
    sessionsDir: string;
    legacyBaseDevDir?: string;
  }) {
    this.config = { appName: options.config?.appName ?? "dormammu" };
    this.baseDevDir = options.baseDevDir;
    this.sessionsDir = options.sessionsDir;
    this.legacyBaseDevDir = options.legacyBaseDevDir ?? options.baseDevDir;
  }

  static normalizeSessionId(value: string): string {
    const normalized = value
      .trim()
      .replace(/[^a-zA-Z0-9._-]+/g, "-")
      .replace(/^[-._]+|[-._]+$/g, "");
    if (!normalized) {
      throw new Error("session_id must contain at least one safe filename character.");
    }
    return normalized;
  }

  async generatedSessionId(timestamp: string): Promise<string> {
    const compact = timestamp
      .replaceAll("-", "")
      .replaceAll(":", "")
      .replaceAll("+", "-")
      .replaceAll("T", "-");
    const base = `${this.config.appName}-${compact}`;
    let candidate = base;
    let sequence = 1;
    while (await exists(path.join(this.sessionsDir, candidate))) {
      candidate = `${base}-${String(sequence).padStart(2, "0")}`;
      sequence += 1;
    }
    return candidate;
  }

  static async currentSessionId(sessionPath: string): Promise<string | null> {
    if (!(await exists(sessionPath))) {
      return null;
    }
    let payload: JsonObject;
    try {
      payload = await readJson(sessionPath);
    } catch (error) {
      if (error instanceof SyntaxError) {
        return null;
      }
      throw error;
    }
    const sessionId = payload.session_id ?? payload.active_session_id;
    return sessionId ? String(sessionId) : null;
  }

  async readActiveSessionId(): Promise<string | null> {
    return SessionManager.currentSessionId(path.join(this.baseDevDir, "session.json"));
  }

  async listSessions(): Promise<SessionSummary[]> {
    const activeSessionId = await this.readActiveSessionId();
    if (!(await exists(this.sessionsDir))) {
      return [];
    }

    const sessions: SessionSummary[] = [];
    for (const entry of await readdir(this.sessionsDir, { withFileTypes: true })) {
      if (!entry.isDirectory()) {
        continue;
      }
      const sessionDir = path.join(this.sessionsDir, entry.name);
      const sessionPath = path.join(sessionDir, "session.json");
      if (!(await exists(sessionPath))) {
        continue;
      }

      const sessionState = await readJson(sessionPath);
      const workflowPath = path.join(sessionDir, "workflow_state.json");
      const workflowState = (await exists(workflowPath)) ? await readJson(workflowPath) : {};
      const bootstrap = isRecord(sessionState.bootstrap) ? sessionState.bootstrap : {};
      const rawGoal = bootstrap.goal ?? sessionState.goal ?? "";
      const goal = String(rawGoal);
      const goalSummary = goal.length > 120 ? `${goal.slice(0, 120)}...` : goal;
      const loopState = isRecord(sessionState.loop)
        ? sessionState.loop
        : isRecord(workflowState.loop)
          ? workflowState.loop
          : {};
      const worktrees = normalizeWorktrees(sessionState.worktrees);
      const sessionId = sessionState.session_id;
      sessions.push({
        session_id: sessionId,
        snapshot_dir: sessionDir,
        created_at: sessionState.created_at,
        updated_at: sessionState.updated_at,
        goal: goalSummary,
        is_active: sessionId === activeSessionId,
        supervisor_verdict: loopState.latest_supervisor_verdict,
        attempts_completed: loopState.attempts_completed,
        active_worktree_id: worktrees.activeWorktreeId,
        managed_worktree_count: worktrees.managedCount
      });
    }

    return sessions.sort((left, right) => {
      const leftKey = String(left.updated_at ?? left.created_at ?? "");
      const rightKey = String(right.updated_at ?? right.created_at ?? "");
      return leftKey.localeCompare(rightKey);
    });
  }

  async hasLegacyRootSnapshot(): Promise<boolean> {
    for (const filename of [...CORE_STATE_FILENAMES, "TASKS.md"]) {
      if (await exists(path.join(this.legacyBaseDevDir, filename))) {
        return true;
      }
    }
    return false;
  }

  static async copyStateSnapshot(sourceDir: string, targetDir: string): Promise<void> {
    await mkdir(targetDir, { recursive: true });
    for (const filename of [...CORE_STATE_FILENAMES, ...OPTIONAL_STATE_FILENAMES]) {
      await copyTextFileIfPresent(path.join(sourceDir, filename), path.join(targetDir, filename));
    }

    const sourceTasksPath = path.join(sourceDir, "TASKS.md");
    const targetTasksPath = path.join(targetDir, "TASKS.md");
    await copyTextFileIfPresent(sourceTasksPath, targetTasksPath);

    const targetPlanPath = path.join(targetDir, "PLAN.md");
    if ((await exists(sourceTasksPath)) && !(await exists(targetPlanPath))) {
      await writeFile(targetPlanPath, await readFile(sourceTasksPath, "utf8"), "utf8");
    }
  }

  async migrateLegacyRootSnapshot(options: { timestamp?: string } = {}): Promise<string | null> {
    const activeSessionId = await this.readActiveSessionId();
    if (
      activeSessionId !== null &&
      (await exists(path.join(this.sessionsDir, activeSessionId)))
    ) {
      return activeSessionId;
    }
    const legacySessionId = await SessionManager.currentSessionId(
      path.join(this.legacyBaseDevDir, "session.json")
    );
    if (
      legacySessionId !== null &&
      (await exists(path.join(this.sessionsDir, legacySessionId)))
    ) {
      return legacySessionId;
    }
    if (!(await this.hasLegacyRootSnapshot())) {
      return null;
    }

    const sessionId =
      legacySessionId ??
      (await this.generatedSessionId(options.timestamp ?? new Date().toISOString()));
    const targetDir = path.join(this.sessionsDir, sessionId);
    await SessionManager.copyStateSnapshot(this.legacyBaseDevDir, targetDir);

    const legacyLogsDir = path.join(this.legacyBaseDevDir, "logs");
    const targetLogsDir = path.join(targetDir, "logs");
    if ((await exists(legacyLogsDir)) && !(await exists(targetLogsDir))) {
      await cp(legacyLogsDir, targetLogsDir, { recursive: true });
    }

    const sessionPath = path.join(targetDir, "session.json");
    if (await exists(sessionPath)) {
      const sessionState = await readJson(sessionPath);
      sessionState.session_id = sessionId;
      sessionState.state_schema_version ??= STATE_SCHEMA_VERSION;
      await writeJson(sessionPath, sessionState);
    }
    return sessionId;
  }

  async readLegacyRootSessionPayload(): Promise<JsonObject | null> {
    const sessionPath = path.join(this.legacyBaseDevDir, "session.json");
    if (!(await exists(sessionPath))) {
      return null;
    }
    const payload = await readJson(sessionPath);
    if ("active_session_id" in payload) {
      return null;
    }
    if (!("session_id" in payload)) {
      return null;
    }
    return payload;
  }

}
